README.md
## WX-ROS2
ROS 2 Interface for Wayties V2X (WX)

### Prerequisites 
ROS 2 Foxy Fitzroy [[ROS2 Foxy Install](https://docs.ros.org/en/foxy/Installation/Linux-Install-Debians.html)], Ubuntu 20.04
###  
### Clone & Build
```
git clone https://github.com/wayties/wx_ros2.git
cd wx_ros2
colcon build --symlink-install
```

### Getting Started 
Open a new terminal, navigate to `ws_ros2`, and source the setup files:
```
source install/local_setup.bash
```

Now run the ROS2 WsmpPoti application: 
```
$ ros2 run wsmp_poti wsmp_poti
[INFO] [1615358883.854047800] [WsmpPoti]: [SUB][FACX-POTI-IND] nts: 1615358883.853653669, seq: 111
[INFO] [1615358883.854164600] [WsmpPoti]:                      Fix Status (valid)  : 3 (0: NO, 1: TIME, 2: 2D FIX, 3: 3D FIX)
[INFO] [1615358883.854234000] [WsmpPoti]:                      Used SV    (invalid): 0 
[INFO] [1615358883.854299600] [WsmpPoti]:                      Latitude   (valid)  :  37.4110107 [deg]
[INFO] [1615358883.854416900] [WsmpPoti]:                      Longitude  (valid)  : 127.0954984 [deg]
[INFO] [1615358883.854442600] [WsmpPoti]:                      Elevation  (valid)  :       112.3 [meter]
[INFO] [1615358883.854563400] [WsmpPoti]:                      Speed      (valid)  :      109.64 [m/s]
[INFO] [1615358883.854702500] [WsmpPoti]: [SUB][FACX-WSMP-IND] nts: 1615358883.854152203, seq: 111
[INFO] [1615358883.854767900] [WsmpPoti]:                      WSM payload size: 46
[INFO] [1615358883.910554900] [WsmpPoti]: [PUB][FACX-WSMP-REQ] nts: 1615358883.910467863, seq: 123
[INFO] [1615358883.910654100] [WsmpPoti]:                      WSM Payload Size: 43
[INFO] [1615358883.910745800] [WsmpPoti]:                      User Priority   : 5
[INFO] [1615358883.910805900] [WsmpPoti]:                      PSID            : 0x7F
[INFO] [1615358883.911364400] [WsmpPoti]: [SUB][FACX-WSMP-CFM] nts: 1615358883.910467863, seq: 123
[INFO] [1615358883.911426700] [WsmpPoti]:                      Result Code: ACCEPTED
```
If there is no device, you can test using the emulator:
```
$ ros2 run wsmp_poti_emu ros2 run wsmp_poti_emu
[INFO] [1615358883.853735500] [WsmpPotiEmu]: [PUB][FACX-POTI-IND] nts: 1615358883.853653669, seq: 111
[INFO] [1615358883.853842300] [WsmpPotiEmu]:                      Fix Status: 3 (0: NO, 1: TIME, 2: 2D FIX, 3: 3D FIX)
[INFO] [1615358883.853875700] [WsmpPotiEmu]:                      Latitude :  37.4110107 [deg]
[INFO] [1615358883.853936900] [WsmpPotiEmu]:                      Longitude: 127.0954984 [deg]
[INFO] [1615358883.853999900] [WsmpPotiEmu]:                      Elevation:       112.3 [meter]
[INFO] [1615358883.854067300] [WsmpPotiEmu]:                      Speed    :      109.64 [m/s]
[INFO] [1615358883.854307400] [WsmpPotiEmu]: [PUB][FACX-WSMP-IND] nts: 1615358883.854152203, seq: 111
[INFO] [1615358883.854421300] [WsmpPotiEmu]:                      WSM payload size: 46
[INFO] [1615358883.910868200] [WsmpPotiEmu]: [SUB][FACX-WSMP-REQ] nts: 1615358883.910467863, seq: 123
[INFO] [1615358883.910949100] [WsmpPotiEmu]:                      WSM payload size: 43
[INFO] [1615358883.910999900] [WsmpPotiEmu]:                      User Priority (valid)  : 5
[INFO] [1615358883.911049100] [WsmpPotiEmu]:                      PSID          (valid)  : 0x7F
[INFO] [1615358883.911132100] [WsmpPotiEmu]: [PUB][FACX-WSMP-CFM] nts: 1615358883.910467863, seq: 123
[INFO] [1615358883.911189100] [WsmpPotiEmu]:                      Result Code: ACCEPTED
```
FACX-POTI-IND and FACX-WSMP-IND are simplex indication. \
But, FACX-WSMP-CFM is a response of FACX-WSMP-REQ.
```
+----------+                           +-----------------+  
|          |  <--- FACX-POTI-IND ----  |                 | 
|          |                           |                 |
|   ROS2   |  <--- FACX-WSMP-CFM ------------     WX     |
|          |                           |    ^   Device   |
| WsmpPoti |                           |    |     or     |  
|          |  ---- FACX-WSMP-REQ ----------->  Emulator  |
|          |                           |                 |
|          |  <--- FACX-WSMP-IND ----  |                 |
+----------+                           +-----------------+
```

### Multiple Devices Handling
When you control multiple wx devices at the same time, \
you need to set a specific domain ID(0 ~ 76) of ROS2 for each wx device. \
[[Configuring your ROS2 environment 3.1 The ROS_DOMAIN_ID variable](https://docs.ros.org/en/foxy/Tutorials/Configuring-ROS2-Environment.html)]

rclcpp/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for rclcpp package of wx_ros2
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.5.0 (2021-04-01)
------------------
* Update enum type and FacxWsmpRequest period


0.4.0 (2021-03-29)
------------------
* Update to extension message for WsmpPoti
* Update to extension message for WsmpPotiEmu 


0.3.0 (2021-03-08)
------------------
* Initial release
* Add basic WsmpPoti application
* Add basic WsmpPotiEmu application


rclcpp/wsmp_poti_emu/src/wsmp_poti_emu.cpp
// Copyright 2020-2021 Wayties, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"

#include "wx_msgs/msg/facx_poti_indication.hpp"

#include "wx_msgs/msg/facx_wsmp_request.hpp"
#include "wx_msgs/msg/facx_wsmp_confirm.hpp"
#include "wx_msgs/msg/facx_wsmp_indication.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;

using namespace wx_msgs::msg;

// The constant topic name is WX protocol. Do not change.
const std::string kTOPIC_FACX_POTI_IND = "wx/facx/poti/ind";
const std::string kTOPIC_FACX_WSMP_REQ = "wx/facx/wsmp/req";
const std::string kTOPIC_FACX_WSMP_CFM = "wx/facx/wsmp/cfm";
const std::string kTOPIC_FACX_WSMP_IND = "wx/facx/wsmp/ind";

enum kWsmpValidity {
  SEC_TYPE                    = (0x01 << 0),
  SIGNER_IDENT_TYPE           = (0x01 << 1),
  SOURCE_MAC_ADDRESS          = (0x01 << 2),
  CHANNEL_IDENTIFIER          = (0x01 << 3),
  TIME_SLOT                   = (0x01 << 4),
  FLOW_SLOT                   = (0x01 << 5),
  DATA_RATE                   = (0x01 << 6),
  TRANSMIT_POWER_LEVEL        = (0x01 << 7),
  USER_PRIORITY               = (0x01 << 8),
  EXPIRY_TIME                 = (0x01 << 9),
  PEER_MAC_ADDRESS            = (0x01 << 10),
  PROVIDER_SERVICE_IDENTIFIER = (0x01 << 11),
};

enum eFixStatus {
  NO,
  TIME,
  FIX2D,  
  FIX3D,  
};

enum kPotiIndValiditiy {
  FIX_STATUS                         = (0x01 << 0),
  FIX_TYPE                           = (0x01 << 1),
  NUM_USED_SV                        = (0x01 << 2),
  NUM_VISIBLE_SV                     = (0x01 << 3),
  PDOP                               = (0x01 << 4), 
  TIMESTAMP                          = (0x01 << 5),
  LATITUDE                           = (0x01 << 6),
  LONGITUDE                          = (0x01 << 7),
  ELEVATION_MSL                      = (0x01 << 8),
  ELEVATION_ELLIPSOID                = (0x01 << 9),
  SPEED_GROUND                       = (0x01 << 10),
  SPEED_VEHICLE                      = (0x01 << 11),
  HEADING_MOTION                     = (0x01 << 12),
  HEADING_VEHICLE                    = (0x01 << 13),
  LONGITUDINAL_ACCELERATION          = (0x01 << 14),
  LATERAL_ACCELERATION               = (0x01 << 15),
  VERTICAL_ACCELERATION              = (0x01 << 16),
  YAW_RATE                           = (0x01 << 17),
  SEMI_MAJOR_AXIS_ACCURACY           = (0x01 << 18),
  SEMI_MINOR_AXIS_ACCURACY           = (0x01 << 19),
  SEMI_MAJOR_ORIENTAtion             = (0x01 << 20),
  TIME_ACCURACY                      = (0x01 << 21),
  LATITUDE_ACCURACY                  = (0x01 << 22),
  LONGITUDE_ACCURACY                 = (0x01 << 23),
  ELEVATION_ACCURACY                 = (0x01 << 24),
  SPEED_ACCURACY                     = (0x01 << 25),
  HEADING_ACCURACY                   = (0x01 << 26),
  LONGITUDINAL_ACCELERATION_ACCURACY = (0x01 << 27),
  LATTERAL_ACCELERATION_ACCURACY     = (0x01 << 28),
  VERTICAL_ACCELERATION_ACCURACY     = (0x01 << 29),
  YAW_RATE_ACCURACY                  = (0x01 << 30),
};

enum eResultCode {
  ACCEPTED,                      
  REJECTED_MAX_LENGTH_EXCCEEDED, 
  REJECTED_UNSPECIFIED,          
};

class WsmpPotiEmu: public rclcpp::Node
{
public:
  WsmpPotiEmu()
  : Node("WsmpPotiEmu"), poti_ind_seq_(0), wsmp_ind_seq_(0)
  {
    // initialize timers, If you want to change current(10Hz) period, change the 100ms 
    poti_ind_timer_ = this->create_wall_timer(100ms, std::bind(&WsmpPotiEmu::on_poti_ind_timer_100ms, this));
    wsmp_ind_timer_ = this->create_wall_timer(100ms, std::bind(&WsmpPotiEmu::on_wsmp_ind_timer_100ms, this));

    // initialize publisher, subscriber
    pub_facx_poti_ind_ = this->create_publisher<FacxPotiIndication>(kTOPIC_FACX_POTI_IND, 10);

    sub_facx_wsmp_req_ = this->create_subscription<FacxWsmpRequest>(
      kTOPIC_FACX_WSMP_REQ, 10, std::bind(&WsmpPotiEmu::facx_wsmp_req, this, _1));

    pub_facx_wsmp_cfm_ = this->create_publisher<FacxWsmpConfirm>(kTOPIC_FACX_WSMP_CFM, 10);

    pub_facx_wsmp_ind_ = this->create_publisher<FacxWsmpIndication>(kTOPIC_FACX_WSMP_IND, 10);
  }

private:
  /**
   * Timer Handler for Fac-POTI.indication
   * 
   * This handler sends a indication with static POTI(Positioning and Timing) information at 10 Hz.
   */
  void on_poti_ind_timer_100ms()
  {
    auto poti = FacxPotiIndication();

    auto now = std::chrono::high_resolution_clock::now();
    uint64_t nsec = std::chrono::duration_cast<std::chrono::nanoseconds>
                    (now.time_since_epoch()).count();
    poti.wx_header.nts = nsec;            
    poti.wx_header.seq = poti_ind_seq_++;

    // set a static FacPotiIndication data
    poti.fix_status          = FIX3D;
    poti.latitude            =  374110107; // 1/10 micro degree, SAE J2735
    poti.longitude           = 1270954984; // 1/10 micro degree, SAE J2735
    poti.elevation_ellipsoid = 1123;       // Units of 10 cm(SAE J2735), Ellipsoid(SAE J2945/1)
    poti.speed_ground        = 5482;       // Units of 0.02 m/s(SAE J2735)

    poti.validity = FIX_STATUS |
                    // NUM_USED_SV | // invalid case
                    LATITUDE |
                    LONGITUDE |
                    ELEVATION_ELLIPSOID |
                    SPEED_GROUND;

    pub_facx_poti_ind_->publish(poti);

    RCLCPP_INFO(this->get_logger(), 
      "[PUB][FACX-POTI-IND] nts: %.9lf, seq: %llu", double(poti.wx_header.nts) * 1e-9, 
                                                   poti.wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     Fix Status: %s (0: NO, 1: TIME, 2: 2D FIX, 3: 3D FIX)", 
      std::to_string(poti.fix_status).c_str());
    RCLCPP_INFO(this->get_logger(), 
      "                     Latitude : %11.7lf [deg]",   (double)poti.latitude  * 1e-7);
    RCLCPP_INFO(this->get_logger(), 
      "                     Longitude: %11.7lf [deg]",   (double)poti.longitude * 1e-7);
    RCLCPP_INFO(this->get_logger(), 
      "                     Elevation: %11.1lf [meter]", (double)poti.elevation_ellipsoid * 0.1);
    RCLCPP_INFO(this->get_logger(), 
      "                     Speed    : %11.2lf [m/s]",   (double)poti.speed_ground * 0.02);
  }
 
  /**
   * Timer Handler for Fac-WSMP.indication
   * 
   * This handler sends a indication with the static WSMP packet at 10 Hz.
   */
  void on_wsmp_ind_timer_100ms()
  {
    auto wsmp = FacxWsmpIndication();

    auto now = std::chrono::high_resolution_clock::now();
    uint64_t nsec = std::chrono::duration_cast<std::chrono::nanoseconds>
                    (now.time_since_epoch()).count();
    wsmp.wx_header.nts = nsec;
    wsmp.wx_header.seq = wsmp_ind_seq_++;

    std::string str("This is a sample payload of FacWsmpIndication.");
    std::vector<uint8_t> data(str.begin(), str.end()); 
    wsmp.data = data;

    pub_facx_wsmp_ind_->publish(wsmp);

    RCLCPP_INFO(this->get_logger(), 
      "[PUB][FACX-WSMP-IND] nts: %.9lf, seq: %llu", (double)wsmp.wx_header.nts * 1e-9, 
                                                   wsmp.wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     WSM payload size: %d", wsmp.data.size());
  }

  /**
   * Callback Function for Fac-WSMP.request
   * 
   * This function receives a request for Fac-WSMP.request and sends a response to it as Fac-WSMP.confirm.
   */
  void facx_wsmp_req(const FacxWsmpRequest::SharedPtr req) 
  {
    RCLCPP_INFO(this->get_logger(), 
      "[SUB][FACX-WSMP-REQ] nts: %.9lf, seq: %llu", (double)req->wx_header.nts * 1e-9, 
                                                   req->wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     WSM payload size: %d", req->data.size());
    RCLCPP_INFO(this->get_logger(), 
      "                     User Priority %s: %d",   
      ((req->validity & USER_PRIORITY) ? "(valid)  " : "(invalid)"),
      ((req->validity & USER_PRIORITY) ? req->user_priority : 0));
    RCLCPP_INFO(this->get_logger(),  
      "                     PSID          %s: 0x%X", 
      ((req->validity & PROVIDER_SERVICE_IDENTIFIER) ? "(valid)  " : "(invalid)"),
      ((req->validity & PROVIDER_SERVICE_IDENTIFIER) ? req->provider_service_identifier : 0));

    auto cfm = FacxWsmpConfirm();

    cfm.wx_header.nts = req->wx_header.nts;
    cfm.wx_header.seq = req->wx_header.seq;

    cfm.result_code = ACCEPTED;

    pub_facx_wsmp_cfm_->publish(cfm);

    RCLCPP_INFO(this->get_logger(), 
      "[PUB][FACX-WSMP-CFM] nts: %.9lf, seq: %llu", double(cfm.wx_header.nts) * 1e-9, 
                                                   cfm.wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     Result Code: ACCEPTED");
  }

  // member variable for POTI indication timer
  rclcpp::TimerBase::SharedPtr poti_ind_timer_;

  // member variable for WSMP indication timer
  rclcpp::TimerBase::SharedPtr wsmp_ind_timer_;

  // member variable for POTI publisher
  rclcpp::Publisher   <FacxPotiIndication>::SharedPtr pub_facx_poti_ind_;

  // member variable for WSMP publisher, subscriber
  rclcpp::Subscription<FacxWsmpRequest>::SharedPtr    sub_facx_wsmp_req_;
  rclcpp::Publisher   <FacxWsmpConfirm>::SharedPtr    pub_facx_wsmp_cfm_;
  rclcpp::Publisher   <FacxWsmpIndication>::SharedPtr pub_facx_wsmp_ind_;

  uint64_t poti_ind_seq_;
  uint64_t wsmp_ind_seq_;
};

int main(int argc, char *argv[])
{
  // ROS2 environment initialize
  rclcpp::init(argc, argv);
  // ROS2 execute main thread
  rclcpp::spin(std::make_shared<WsmpPotiEmu>());
  // ROs2 shutdown main thread 
  rclcpp::shutdown(); 

  return EXIT_SUCCESS;
}


rclcpp/wsmp_poti/src/wsmp_poti.cpp
// Copyright 2020-2021 Wayties, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <string>
#include <chrono>
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"

#include "wx_msgs/msg/facx_poti_indication.hpp"

#include "wx_msgs/msg/facx_wsmp_request.hpp"
#include "wx_msgs/msg/facx_wsmp_confirm.hpp"
#include "wx_msgs/msg/facx_wsmp_indication.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;

using namespace wx_msgs::msg;

// The constant topic name is WX protocol. Do not change.
const std::string kTOPIC_FACX_POTI_IND = "wx/facx/poti/ind";
const std::string kTOPIC_FACX_WSMP_REQ = "wx/facx/wsmp/req";
const std::string kTOPIC_FACX_WSMP_CFM = "wx/facx/wsmp/cfm";
const std::string kTOPIC_FACX_WSMP_IND = "wx/facx/wsmp/ind";

enum kWsmpValidity {
  SEC_TYPE                    = (0x01 << 0),
  SIGNER_IDENT_TYPE           = (0x01 << 1),
  SOURCE_MAC_ADDRESS          = (0x01 << 2),
  CHANNEL_IDENTIFIER          = (0x01 << 3),
  TIME_SLOT                   = (0x01 << 4),
  FLOW_SLOT                   = (0x01 << 5),
  DATA_RATE                   = (0x01 << 6),
  TRANSMIT_POWER_LEVEL        = (0x01 << 7),
  USER_PRIORITY               = (0x01 << 8),
  EXPIRY_TIME                 = (0x01 << 9),
  PEER_MAC_ADDRESS            = (0x01 << 10),
  PROVIDER_SERVICE_IDENTIFIER = (0x01 << 11),
};

enum eFixStatus {
  NO,
  TIME,
  FIX2D,  
  FIX3D,  
};

enum kPotiIndValiditiy {
  FIX_STATUS                         = (0x01 << 0),
  FIX_TYPE                           = (0x01 << 1),
  NUM_USED_SV                        = (0x01 << 2),
  NUM_VISIBLE_SV                     = (0x01 << 3),
  PDOP                               = (0x01 << 4), 
  TIMESTAMP                          = (0x01 << 5),
  LATITUDE                           = (0x01 << 6),
  LONGITUDE                          = (0x01 << 7),
  ELEVATION_MSL                      = (0x01 << 8),
  ELEVATION_ELLIPSOID                = (0x01 << 9),
  SPEED_GROUND                       = (0x01 << 10),
  SPEED_VEHICLE                      = (0x01 << 11),
  HEADING_MOTION                     = (0x01 << 12),
  HEADING_VEHICLE                    = (0x01 << 13),
  LONGITUDINAL_ACCELERATION          = (0x01 << 14),
  LATERAL_ACCELERATION               = (0x01 << 15),
  VERTICAL_ACCELERATION              = (0x01 << 16),
  YAW_RATE                           = (0x01 << 17),
  SEMI_MAJOR_AXIS_ACCURACY           = (0x01 << 18),
  SEMI_MINOR_AXIS_ACCURACY           = (0x01 << 19),
  SEMI_MAJOR_ORIENTAtion             = (0x01 << 20),
  TIME_ACCURACY                      = (0x01 << 21),
  LATITUDE_ACCURACY                  = (0x01 << 22),
  LONGITUDE_ACCURACY                 = (0x01 << 23),
  ELEVATION_ACCURACY                 = (0x01 << 24),
  SPEED_ACCURACY                     = (0x01 << 25),
  HEADING_ACCURACY                   = (0x01 << 26),
  LONGITUDINAL_ACCELERATION_ACCURACY = (0x01 << 27),
  LATTERAL_ACCELERATION_ACCURACY     = (0x01 << 28),
  VERTICAL_ACCELERATION_ACCURACY     = (0x01 << 29),
  YAW_RATE_ACCURACY                  = (0x01 << 30),
};

enum eResultCode {
  ACCEPTED,                      
  REJECTED_MAX_LENGTH_EXCCEEDED, 
  REJECTED_UNSPECIFIED,          
};

class WsmpPoti: public rclcpp::Node
{
public:
  WsmpPoti()
  : Node("WsmpPoti"), seq_(0)
  {
    // initialize timers, If you want to change current(10Hz) period, change the 100ms 
    timer_ = this->create_wall_timer(100ms, std::bind(&WsmpPoti::on_timer_100ms, this));

    // initialize publisher, subscriber
    sub_facx_poti_ind_ = this->create_subscription<FacxPotiIndication>(
      kTOPIC_FACX_POTI_IND, 10, std::bind(&WsmpPoti::facx_poti_ind, this, _1));

    pub_facx_wsmp_req_ = this->create_publisher<FacxWsmpRequest>(kTOPIC_FACX_WSMP_REQ, 10);

    sub_facx_wsmp_cfm_ = this->create_subscription<FacxWsmpConfirm>(
      kTOPIC_FACX_WSMP_CFM, 10, std::bind(&WsmpPoti::facx_wsmp_cfm, this, _1));

    sub_facx_wsmp_ind_ = this->create_subscription<FacxWsmpIndication>(
      kTOPIC_FACX_WSMP_IND, 10, std::bind(&WsmpPoti::facx_wsmp_ind, this, _1));
  }

private:
  /**
   * Timer Handler for Fac-WSMP.request
   * 
   * This handler sends a request to transmit WSM at 10 Hz to device.
   */
  void on_timer_100ms()
  {
    auto req = FacxWsmpRequest();

    auto now = std::chrono::high_resolution_clock::now();
    uint64_t nsec = std::chrono::duration_cast<std::chrono::nanoseconds>
                    (now.time_since_epoch()).count();
    req.wx_header.nts = nsec; 
    req.wx_header.seq = seq_++; 

    req.user_priority               = 5;    // no ciritical event flags, SAE J2945/1
    req.provider_service_identifier = 0x7F; // 127

    std::string str("This is a sample payload of FacWsmpRequest.");
    std::vector<uint8_t> data(str.begin(), str.end()); 
    req.data = data;

    req.validity = USER_PRIORITY |
                   PROVIDER_SERVICE_IDENTIFIER;

    pub_facx_wsmp_req_->publish(req);

    RCLCPP_INFO(this->get_logger(), 
      "[PUB][FACX-WSMP-REQ] nts: %.9lf, seq: %llu", double(req.wx_header.nts) * 1e-9, 
                                                   req.wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     WSM Payload Size: %d",   req.data.size());
    RCLCPP_INFO(this->get_logger(), 
      "                     User Priority   : %d",   req.user_priority);
    RCLCPP_INFO(this->get_logger(),  
      "                     PSID            : 0x%X", req.provider_service_identifier);
  }

  /**
   * Callback Function for Fac-POTI.indication
   * 
   * This function receives POTI(Positioning and Timing) information from device. 
   */
  void facx_poti_ind(const FacxPotiIndication::SharedPtr ind) const
  {
    RCLCPP_INFO(this->get_logger(), 
      "[SUB][FACX-POTI-IND] nts: %.9lf, seq: %llu", double(ind->wx_header.nts) * 1e-9, 
                                                   ind->wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     Fix Status %s: %d (0: NO, 1: TIME, 2: 2D FIX, 3: 3D FIX)",
      ((ind->validity & FIX_STATUS) ? "(valid)  " : "(invalid)"),
      ((ind->validity & FIX_STATUS) ? ind->fix_status : 0));
    RCLCPP_INFO(this->get_logger(), 
      "                     Used SV    %s: %d ",
      ((ind->validity & NUM_USED_SV) ? "(valid)  " : "(invalid)"),
      ((ind->validity & NUM_USED_SV) ? ind->num_used_sv: 0));
    RCLCPP_INFO(this->get_logger(), 
      "                     Latitude   %s: %11.7lf [deg]",   // degree
      ((ind->validity & LATITUDE) ? "(valid)  " : "(invalid)"),
      ((ind->validity & LATITUDE) ? (double)ind->latitude  * 1e-7 : 0.00));
    RCLCPP_INFO(this->get_logger(), 
      "                     Longitude  %s: %11.7lf [deg]",   // degree
      ((ind->validity & LONGITUDE) ? "(valid)  " : "(invalid)"),
      ((ind->validity & LONGITUDE) ? (double)ind->longitude * 1e-7 : 0.00));
    RCLCPP_INFO(this->get_logger(), 
      "                     Elevation  %s: %11.1lf [meter]", // meter 
      ((ind->validity & ELEVATION_ELLIPSOID) ? "(valid)  " : "(invalid)"),
      ((ind->validity & ELEVATION_ELLIPSOID) ? (double)ind->elevation_ellipsoid * 0.1 : 0.00));
    RCLCPP_INFO(this->get_logger(), 
      "                     Speed      %s: %11.2lf [m/s]",   // m/s 
      ((ind->validity & SPEED_GROUND) ? "(valid)  " : "(invalid)"),
      ((ind->validity & SPEED_GROUND) ? (double)ind->speed_ground * 0.02 : 0.00));
  }

  /**
   * Callback Function for Fac-WSMP.confirm
   * 
   * This function receives response for Fac-WSMP.request from device. 
   */
  void facx_wsmp_cfm(const FacxWsmpConfirm::SharedPtr cfm) const
  {
    RCLCPP_INFO(this->get_logger(), 
      "[SUB][FACX-WSMP-CFM] nts: %.9lf, seq: %llu", double(cfm->wx_header.nts) * 1e-9, 
                                                   cfm->wx_header.seq);
    if (cfm->result_code == ACCEPTED) {
      RCLCPP_INFO(this->get_logger(), 
      "                     Result Code: ACCEPTED");
    }
    else if (cfm->result_code == REJECTED_MAX_LENGTH_EXCCEEDED) {
      RCLCPP_INFO(this->get_logger(), 
      "                     Result Code: REJECTED_MAX_LENGTH_EXCCEEDED");
    }
    else if (cfm->result_code == REJECTED_UNSPECIFIED) {
      RCLCPP_INFO(this->get_logger(), 
      "                     Result Code: REJECTED_UNSPECIFIED");
    }
    else {
      RCLCPP_INFO(this->get_logger(), 
      "                     Result Code: UNKNOWN");
    }
  }

  /**
   * Callback Function for Fac-WSMP.indication
   * 
   * This function receives WSM packet as Fac-WSMP.indication from device.
   */
  void facx_wsmp_ind(const FacxWsmpIndication::SharedPtr ind) const
  {
    RCLCPP_INFO(this->get_logger(), 
      "[SUB][FACX-WSMP-IND] nts: %.9lf, seq: %llu", double(ind->wx_header.nts) * 1e-9, 
                                                   ind->wx_header.seq);
    RCLCPP_INFO(this->get_logger(), 
      "                     WSM payload size: %d", ind->data.size());
  }

  // member variable for WSMP request timer
  rclcpp::TimerBase::SharedPtr timer_;

  // member variable for POTI publisher
  rclcpp::Subscription<FacxPotiIndication>::SharedPtr sub_facx_poti_ind_;

  // member variable for WSMP publisher, subscriber
  rclcpp::Publisher   <FacxWsmpRequest>::SharedPtr    pub_facx_wsmp_req_;
  rclcpp::Subscription<FacxWsmpIndication>::SharedPtr sub_facx_wsmp_ind_;
  rclcpp::Subscription<FacxWsmpConfirm>::SharedPtr    sub_facx_wsmp_cfm_;

  uint64_t seq_;
};

int main(int argc, char *argv[])
{
  // ROS2 environment initialize
  rclcpp::init(argc, argv);
  // ROS2 execute main thread
  rclcpp::spin(std::make_shared<WsmpPoti>());
  // ROs2 shutdown main thread 
  rclcpp::shutdown(); 

  return EXIT_SUCCESS;
}
